cmd/root.go

```diff
@@ -3,242 +3,237 @@ package cmd
 import (
 	"fmt"
 
-	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
-	"github.com/spf13/viper"
 
 	"github.com/spacemeshos/go-spacemesh/common/types"
 	"github.com/spacemeshos/go-spacemesh/config"
 	"github.com/spacemeshos/go-spacemesh/config/presets"
 	"github.com/spacemeshos/go-spacemesh/node/flags"
 )
 
-var cfg = config.MainnetConfig()
+func AddFlags(flagSet *pflag.FlagSet, cfg *config.Config) (configPath *string) {
```

Comment:
I find named returns for longer functions quite confusing; the return references the variable again anyway, so why not make it just a `(string, err)`?

```suggestion
func AddFlags(flagSet *pflag.FlagSet, cfg *config.Config) (string, err) {
```

Assign config as a value:

```go
configPath := *flagSet.StringP(...)
```

Then in the good case you can `return configPath, nil` and in the bad case `return "", err` instead of panicking 🙂

Reply:
I used a named return to make it obvious what is returned (a pointer to a variable containing a path to config) because it is not obvious by looking at the function _name_. How would you tell what this function returns by looking at:

```go
func AddFlags(flagSet *pflag.FlagSet, cfg *config.Config) (string, err) {
```
? You would need to look through its source code. Alternatively, I could add a comment but I find code superior to comments which often get outdated soon.


> Assign config as a value:
```go
configPath := *flagSet.StringP(...)
```
Then in the good case you can return `configPath, nil` and in the bad case `return "", err` instead of panicking 🙂

I cannot return `string` as the flags were not parsed yet. They are just recorded in the flagset so it knows about them when parsing CLI later. The returned `*string` is a pointer to a variable which will hold the config path _after_ CLI arguments are parsed.

Reply:
But just giving the return value a name doesn't solve the issues you stated? Without reading through the code I don't know that `*configPath` does not yet contain the path to the config file and will change later after a call to a different function.

The point with the error still stands, if we can avoid `panic` I think we should.

Reply:
>Without reading through the code I don't know that *configPath does not yet contain the path to the config file and will change later after a call to a different function.

You do know this if you understand that it **Adds** flags, not **parses** them (this is present in the function name). Reading the code won't help if you don't understand that _adding_ a flag doesn't immediately do the parsing anyway :shrug:.

>The point with the error still stands, if we can avoid panic I think we should.

What panic? Do you mean this the following?
```go
if err := flagSet.MarkHidden("poet-server"); err != nil {
	panic(err) // unreachable
}
```

IMHO the panic is the way to go here. It will **only** fail if `the "poet-server"` flag is removed which is clearly a simple programmer error that is fine to handle by aborting immediately (it must never happen) to catch the mistake asap.